GCS repository does not use new handlers for failed request for every new request

Google Cloud Support was contacted by a customer who sees many 503 errors when doing snapshots to Google Cloud Storage (GCS). We believe to have tracked down the bug to an incorrect initialization of HttpRequest in [GoogleCloudStorageService](https://github.com/elastic/elasticsearch/blob/master/plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageService.java).

The example code in https://cloud.google.com/storage/transfer/create-client#retry attaches new instances of HttpUnsuccessfulResponseHandler and HttpBackOffIOExceptionHandler to every request. However the plugin initializes only one instance of each class in the constructor (line 142-143) and attaches the same instances to every request.

Please also consult the JavaDoc of both handler classes:

* [HttpBackOffUnsuccessfulResponseHandler](https://developers.google.com/api-client-library/java/google-http-java-client/reference/1.20.0/com/google/api/client/http/HttpBackOffUnsuccessfulResponseHandler): "you MUST create a new instance of HttpBackOffIOExceptionHandler with a new instance of BackOff for each instance of HttpRequest."
* [HttpBackOffIOExceptionHandler](https://developers.google.com/api-client-library/java/google-http-java-client/reference/1.20.0/com/google/api/client/http/HttpBackOffIOExceptionHandler): "you MUST create a new instance of HttpBackOffIOExceptionHandler with a new instance of BackOff for each instance of HttpRequest."

We believe that the reuse of the failure handlers causes the client to not retry failed requests anymore after the number of possible backoffs (or backoff time) has been exhausted once for each failure handler. Instead clients immediately fail. This is even more problematic since the snapshot logic in an ES cluster causes an immediate spike of write requests and is thus very prone to temporary failure.

Please move the initialization of the failure handlers from the constructor of `DefaultHttpRequestInitializer` into the method `initialize`

Thank you for your consideration!

**Elasticsearch version** (`bin/elasticsearch --version`): 5.4.3

** plugins:
analysis-icu
analysis-smartcn
repository-gcs
repository-s3

** details for GCS plugin:
Name: repository-gcs
Description: The GCS repository plugin adds Google Cloud Storage support for repositories.
Version: 5.4.3
Native Controller: false
 * Classname: org.elasticsearch.repositories.gcs.GoogleCloudStoragePlugin

** OS: Debian 3.16.43-2+deb8u2

**JVM version** (`java -version`): 1.8.0_131

**Description of the problem including expected versus actual behavior**:

**Steps to reproduce**:

Please include a *minimal* but *complete* recreation of the problem, including
(e.g.) index creation, mappings, settings, query etc.  The easier you make for
us to reproduce it, the more likely that somebody will take the time to look at it.

 1. Use the GCS snapshot repository for multiple snapshots, without restarting in an ES cluster of more than 6 GCE VMs with an index that takes around 30-40s for a snapshot.

**Provide logs (if relevant)**:

```
logs snippet from search-es-data-b-scale-1709191530 VM Instance:
[2017-10-10T03:30:03,139][WARN ][o.e.s.SnapshotShardsService] [search-es-data-b-scale-1709191530] [[listings_v7][14]] [gcs-search-es-snapshots:search-es-10102017_1129_29688/Hy6gbA22Q0WVv9i8KL9PvQ] failed to create snapshot
org.elasticsearch.index.snapshots.IndexShardSnapshotFailedException: Failed to perform snapshot (index files)
    at org.elasticsearch.repositories.blobstore.BlobStoreRepository$SnapshotContext.snapshot(BlobStoreRepository.java:1376) ~[elasticsearch-5.4.3.jar:5.4.3]
    at org.elasticsearch.repositories.blobstore.BlobStoreRepository.snapshotShard(BlobStoreRepository.java:971) ~[elasticsearch-5.4.3.jar:5.4.3]
    at org.elasticsearch.snapshots.SnapshotShardsService.snapshot(SnapshotShardsService.java:382) ~[elasticsearch-5.4.3.jar:5.4.3]
    at org.elasticsearch.snapshots.SnapshotShardsService.access$200(SnapshotShardsService.java:88) ~[elasticsearch-5.4.3.jar:5.4.3]
    at org.elasticsearch.snapshots.SnapshotShardsService$1.doRun(SnapshotShardsService.java:335) [elasticsearch-5.4.3.jar:5.4.3]
    at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:638) [elasticsearch-5.4.3.jar:5.4.3]
    at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) [elasticsearch-5.4.3.jar:5.4.3]
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [?:1.8.0_131]
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [?:1.8.0_131]
    at java.lang.Thread.run(Thread.java:748) [?:1.8.0_131]
Caused by: com.google.api.client.googleapis.json.GoogleJsonResponseException: 503 Service Unavailable
Service Unavailable
    at com.google.api.client.googleapis.json.GoogleJsonResponseException.from(GoogleJsonResponseException.java:145) ~[?:?]
    at com.google.api.client.googleapis.services.json.AbstractGoogleJsonClientRequest.newExceptionOnError(AbstractGoogleJsonClientRequest.java:113) ~[?:?]
    at com.google.api.client.googleapis.services.json.AbstractGoogleJsonClientRequest.newExceptionOnError(AbstractGoogleJsonClientRequest.java:40) ~[?:?]
    at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed(AbstractGoogleClientRequest.java:432) ~[?:?]
    at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed(AbstractGoogleClientRequest.java:352) ~[?:?]
    at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.execute(AbstractGoogleClientRequest.java:469) ~[?:?]
    at org.elasticsearch.repositories.gcs.GoogleCloudStorageBlobStore.lambda$writeBlob$5(GoogleCloudStorageBlobStore.java:219) ~[?:?]
    at java.security.AccessController.doPrivileged(Native Method) ~[?:1.8.0_131]
    at org.elasticsearch.repositories.gcs.GoogleCloudStorageBlobStore.doPrivileged(GoogleCloudStorageBlobStore.java:333) ~[?:?]
    at org.elasticsearch.repositories.gcs.GoogleCloudStorageBlobStore.writeBlob(GoogleCloudStorageBlobStore.java:213) ~[?:?]
    at org.elasticsearch.repositories.gcs.GoogleCloudStorageBlobContainer.writeBlob(GoogleCloudStorageBlobContainer.java:72) ~[?:?]
    at org.elasticsearch.repositories.blobstore.BlobStoreRepository$SnapshotContext.snapshotFile(BlobStoreRepository.java:1432) ~[elasticsearch-5.4.3.jar:5.4.3]
    at org.elasticsearch.repositories.blobstore.BlobStoreRepository$SnapshotContext.snapshot(BlobStoreRepository.java:1374) ~[elasticsearch-5.4.3.jar:5.4.3]
    ... 9 more
```

--- a/plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageService.java
+++ b/plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageService.java
@@ -23,7 +23,6 @@ import com.google.api.client.googleapis.auth.oauth2.GoogleCredential;
 import com.google.api.client.googleapis.javanet.GoogleNetHttpTransport;
 import com.google.api.client.http.HttpBackOffIOExceptionHandler;
 import com.google.api.client.http.HttpBackOffUnsuccessfulResponseHandler;
-import com.google.api.client.http.HttpIOExceptionHandler;
 import com.google.api.client.http.HttpRequest;
 import com.google.api.client.http.HttpRequestInitializer;
 import com.google.api.client.http.HttpUnsuccessfulResponseHandler;
@@ -34,9 +33,7 @@ import com.google.api.services.storage.Storage;
 import com.google.api.services.storage.StorageScopes;
 import org.elasticsearch.ElasticsearchException;
 import org.elasticsearch.common.component.AbstractComponent;
-import org.elasticsearch.common.inject.Inject;
 import org.elasticsearch.common.settings.SecureSetting;
-import org.elasticsearch.common.settings.SecureString;
 import org.elasticsearch.common.settings.Setting;
 import org.elasticsearch.common.settings.Settings;
 import org.elasticsearch.common.unit.TimeValue;
@@ -45,8 +42,6 @@ import org.elasticsearch.env.Environment;
 import java.io.IOException;
 import java.io.InputStream;
 import java.io.UncheckedIOException;
-import java.nio.file.Files;
-import java.nio.file.Path;
 import java.util.Collections;
 import java.util.HashMap;
 import java.util.Map;
@@ -77,16 +72,11 @@ interface GoogleCloudStorageService {
      */
     class InternalGoogleCloudStorageService extends AbstractComponent implements GoogleCloudStorageService {
 
-        private static final String DEFAULT = "_default_";
-
-        private final Environment environment;
-
         /** Credentials identified by client name. */
         private final Map<String, GoogleCredential> credentials;
 
         InternalGoogleCloudStorageService(Environment environment, Map<String, GoogleCredential> credentials) {
             super(environment.settings());
-            this.environment = environment;
             this.credentials = credentials;
         }
 
@@ -132,15 +122,11 @@ interface GoogleCloudStorageService {
             private final TimeValue connectTimeout;
             private final TimeValue readTimeout;
             private final GoogleCredential credential;
-            private final HttpUnsuccessfulResponseHandler handler;
-            private final HttpIOExceptionHandler ioHandler;
 
             DefaultHttpRequestInitializer(GoogleCredential credential, TimeValue connectTimeout, TimeValue readTimeout) {
                 this.credential = credential;
                 this.connectTimeout = connectTimeout;
                 this.readTimeout = readTimeout;
-                this.handler = new HttpBackOffUnsuccessfulResponseHandler(newBackOff());
-                this.ioHandler = new HttpBackOffIOExceptionHandler(newBackOff());
             }
 
             @Override
@@ -152,13 +138,14 @@ interface GoogleCloudStorageService {
                     request.setReadTimeout((int) readTimeout.millis());
                 }
 
-                request.setIOExceptionHandler(ioHandler);
+                request.setIOExceptionHandler(new HttpBackOffIOExceptionHandler(newBackOff()));
                 request.setInterceptor(credential);
 
+                final HttpUnsuccessfulResponseHandler handler = new HttpBackOffUnsuccessfulResponseHandler(newBackOff());
                 request.setUnsuccessfulResponseHandler((req, resp, supportsRetry) -> {
-                            // Let the credential handle the response. If it failed, we rely on our backoff handler
-                            return credential.handleResponse(req, resp, supportsRetry) || handler.handleResponse(req, resp, supportsRetry);
-                        }
+                        // Let the credential handle the response. If it failed, we rely on our backoff handler
+                        return credential.handleResponse(req, resp, supportsRetry) || handler.handleResponse(req, resp, supportsRetry);
+                    }
                 );
             }
 

--- a/plugins/repository-gcs/src/test/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageServiceTests.java
+++ b/plugins/repository-gcs/src/test/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageServiceTests.java
@@ -19,17 +19,35 @@
 
 package org.elasticsearch.repositories.gcs;
 
+import com.google.api.client.googleapis.auth.oauth2.GoogleCredential;
+import com.google.api.client.http.GenericUrl;
+import com.google.api.client.http.HttpIOExceptionHandler;
+import com.google.api.client.http.HttpRequest;
+import com.google.api.client.http.HttpRequestFactory;
+import com.google.api.client.http.HttpRequestInitializer;
+import com.google.api.client.http.HttpResponse;
+import com.google.api.client.http.HttpUnsuccessfulResponseHandler;
+import com.google.api.client.testing.http.MockHttpTransport;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.env.Environment;
+import org.elasticsearch.env.TestEnvironment;
+import org.elasticsearch.repositories.gcs.GoogleCloudStorageService.InternalGoogleCloudStorageService;
+import org.elasticsearch.test.ESTestCase;
+
 import java.io.IOException;
 import java.io.InputStream;
 import java.util.Collections;
 import java.util.Map;
 
-import com.google.api.client.googleapis.auth.oauth2.GoogleCredential;
-import org.elasticsearch.common.settings.Settings;
-import org.elasticsearch.env.Environment;
-import org.elasticsearch.env.TestEnvironment;
-import org.elasticsearch.repositories.gcs.GoogleCloudStorageService.InternalGoogleCloudStorageService;
-import org.elasticsearch.test.ESTestCase;
+import static java.util.Collections.emptyMap;
+import static java.util.Collections.singletonMap;
+import static org.mockito.Matchers.any;
+import static org.mockito.Matchers.anyBoolean;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
 
 public class GoogleCloudStorageServiceTests extends ESTestCase {
 
@@ -47,13 +65,56 @@ public class GoogleCloudStorageServiceTests extends ESTestCase {
             }
         };
         assertSame(cred, service.getCredential("default"));
+
+        service.new DefaultHttpRequestInitializer(cred, null, null);
     }
 
     public void testClientCredential() throws Exception {
         GoogleCredential cred = GoogleCredential.fromStream(getDummyCredentialStream());
-        Map<String, GoogleCredential> credentials = Collections.singletonMap("clientname", cred);
+        Map<String, GoogleCredential> credentials = singletonMap("clientname", cred);
         Environment env = TestEnvironment.newEnvironment(Settings.builder().put("path.home", createTempDir()).build());
         InternalGoogleCloudStorageService service = new InternalGoogleCloudStorageService(env, credentials);
         assertSame(cred, service.getCredential("clientname"));
+    }
+
+    /**
+     * Test that the {@link InternalGoogleCloudStorageService.DefaultHttpRequestInitializer} attaches new instances
+     * of {@link HttpIOExceptionHandler} and {@link HttpUnsuccessfulResponseHandler} for every HTTP requests.
+     */
+    public void testDefaultHttpRequestInitializer() throws IOException {
+        final Environment environment = mock(Environment.class);
+        when(environment.settings()).thenReturn(Settings.EMPTY);
+
+        final GoogleCredential credential = mock(GoogleCredential.class);
+        when(credential.handleResponse(any(HttpRequest.class), any(HttpResponse.class), anyBoolean())).thenReturn(false);
+
+        final TimeValue readTimeout = TimeValue.timeValueSeconds(randomIntBetween(1, 120));
+        final TimeValue connectTimeout = TimeValue.timeValueSeconds(randomIntBetween(1, 120));
+
+        final InternalGoogleCloudStorageService service = new InternalGoogleCloudStorageService(environment, emptyMap());
+        final HttpRequestInitializer initializer = service.new DefaultHttpRequestInitializer(credential, connectTimeout, readTimeout);
+        final HttpRequestFactory requestFactory = new MockHttpTransport().createRequestFactory(initializer);
+
+        final HttpRequest request1 = requestFactory.buildGetRequest(new GenericUrl());
+        assertEquals((int) connectTimeout.millis(), request1.getConnectTimeout());
+        assertEquals((int) readTimeout.millis(), request1.getReadTimeout());
+        assertSame(credential, request1.getInterceptor());
+        assertNotNull(request1.getIOExceptionHandler());
+        assertNotNull(request1.getUnsuccessfulResponseHandler());
+
+        final HttpRequest request2 = requestFactory.buildGetRequest(new GenericUrl());
+        assertEquals((int) connectTimeout.millis(), request2.getConnectTimeout());
+        assertEquals((int) readTimeout.millis(), request2.getReadTimeout());
+        assertSame(request1.getInterceptor(), request2.getInterceptor());
+        assertNotNull(request2.getIOExceptionHandler());
+        assertNotSame(request1.getIOExceptionHandler(), request2.getIOExceptionHandler());
+        assertNotNull(request2.getUnsuccessfulResponseHandler());
+        assertNotSame(request1.getUnsuccessfulResponseHandler(), request2.getUnsuccessfulResponseHandler());
+
+        request1.getUnsuccessfulResponseHandler().handleResponse(null, null, false);
+        verify(credential, times(1)).handleResponse(any(HttpRequest.class), any(HttpResponse.class), anyBoolean());
+
+        request2.getUnsuccessfulResponseHandler().handleResponse(null, null, false);
+        verify(credential, times(2)).handleResponse(any(HttpRequest.class), any(HttpResponse.class), anyBoolean());
     }
 }

